onnx/defs/nn/defs.cc

```diff
@@ -1718,10 +1749,102 @@ ONNX_OPERATOR_SET_SCHEMA(
             "T",
             {"tensor(float16)", "tensor(float)", "tensor(double)"},
             "Constrain input and output types to float tensors.")
+        .TypeConstraint(
+            "T1",
+            {"tensor(bool)"},
+            "Constrain input training_mode to boolean tensors.")
         .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
           propagateShapeAndTypeFromFirstInput(ctx);
-          // TODO in training mode, it may be possible to infer some of
-          // the other outputs as well.
+          if (hasInputShape(ctx, 0)) {
+            propagateShapeFromInputToOutput(ctx, 0, 0);
+            auto& x_input_shape = getInputShape(ctx, 0);
+            int num_channels = 1;
+            if (static_cast<int>(x_input_shape.dim_size()) > 1 &&
+                x_input_shape.dim(1).has_dim_value()) {
```

Comment:
Should handle dim_param for the channel index ?

Reply:
Agree, we need to handle unknown dimensions carefully. The utility function ```unifyInputDim``` handles all cases carefully and will help simplify the code. See https://github.com/onnx/onnx/blob/490895703c82f68e723a8b308369cac4ed0abc0c/onnx/defs/object_detection/defs.cc#L117 for an example of how it is used.

Reply:
Roughly speaking, I think something like below should work (I haven't checked if the input-numbers and dim-numbers are the correct ones):
```cpp
   Dim C;
   unifyInputDim(ctx, 0, 1, C);
   unifyInputDim(ctx, 1, 0, C);
   unifyInputDim(ctx, 2, 0, C);
   etc.
```

Reply:
Basically call ```unifyInputDim``` for every input/dim that should have the same value C

Reply:
And you can use C to set the corresponding output dimension. This will pick up a known-value from any of the input shapes if at least one of the input shapes has a known value for that dimension.